lib/misc/sw-template.js

```diff
@@ -283,12 +265,87 @@ function WebpackServiceWorker(params) {
         return response;
       });
 
-      if (navigateFallbackURL && isNavigateRequest(event.request)) {
-        return handleNavigateFallback(fetching);
+      return fetching;
+    });
+  }
+
+  function networkFirst(event, urlString) {
+    return fetch(event.request).then(function (response) {
+      console.log('[SW]:', response);
+      if (response.ok) {
+        if (DEBUG) {
+          console.log('[SW]:', 'URL [' + urlString + '] from network');
+        }
+
+        return response;
       }
 
-      return fetching;
+      // throw to reach the code in the catch below
+      throw new Error("response is not ok");
+    })
+    // this needs to be in a catch() and not just in the then() above
+    // cause if your network is down, the fetch() will throw
+    ['catch'](function () {
+      if (DEBUG) {
+        console.log('[SW]:', 'URL [' + urlString + '] from cache if possible');
+      }
+
+      return cachesMatch(event.request, CACHE_NAME);
     });
+  }
+
+  self.addEventListener('fetch', function (event) {
+    var requestUrl = event.request.url;
+    var url = new URL(requestUrl);
+    var urlString = undefined;
+
+    if (externals.indexOf(requestUrl) !== -1) {
+      urlString = requestUrl;
+    } else {
+      url.search = '';
+      urlString = url.toString();
+    }
+
+    // Handle only GET requests
+    var isGET = event.request.method === 'GET';
+    var assetMatches = allAssets.indexOf(urlString) !== -1;
+
+    if (!assetMatches && isGET) {
+      // If isn't a cached asset and is a navigation request,
+      // fallback to navigateFallbackURL if available
+      if (navigateFallbackURL && isNavigateRequest(event.request)) {
+        event.respondWith(handleNavigateFallback(fetch(event.request)));
+
+        return;
+      }
+    }
+
+    if (!assetMatches || !isGET) {
+      // Fix for https://twitter.com/wanderview/status/696819243262873600
+      if (url.origin !== location.origin && navigator.userAgent.indexOf('Firefox/44.') !== -1) {
+        event.respondWith(fetch(event.request));
+      }
+
+      return;
+    }
+
+    // Logic of caching / fetching is here
+    // * urlString -- url to match from the CACHE_NAME
+    // * event.request -- original Request to perform fetch() if necessary
+    var resource = undefined;
+
+    if (responseStrategy === "network-first") {
+      resource = networkFirst(event, urlString);
+    }
+    //"cache-first"
+    // (responseStrategy has been validated before)
+    else {
+        resource = cacheFirst(event, urlString);
+      }
+
+    if (navigateFallbackURL && isNavigateRequest(event.request)) {
+      resource = handleNavigateFallback(resource);
+    }
```

Comment:
looks like this block should be handled differently for cache first and network first.
With network first, this can make a weird UX when you get your html page from cache with no css until it's loaded. So flash of unstyled content, which sucks. Any idea how we can improve that?